### PR TITLE
EVM: JSON-RPC endpoints compatibility testing: `eth_getLogs` and `eth_feeHistory`

### PR DESCRIPTION
**Pull request by citizen-stig**
# Summary

---

<a href="https://app.devin.ai/review/sovereign-labs/sovereign-sdk/pull/2387">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

Adds extensive tests coverage for RPC endpoints `eth_getLogs` and `eth_feeHistory`

## Key Changes

1. Fee History Fixes (fee_history.rs)
  - Chain state alignment for next block's base fee
  - Block validation against rollup height
  - Empty percentiles handling per spec
  - Accurate block count when chain is young
2. Effective Gas Price Fix (mod.rs)
  - Fixed receipt effective_gas_price to use proper EIP-1559 formula instead of just returning base fee
3. Test Suites (~4,100 lines)
  - evm_fee_history.rs - Block count, percentiles, pending blocks, gas ratios
  - evm_logs.rs - Block ranges, tags, address/topic filtering, pending logs
4. Test Infrastructure
  - 120s timeout for receipt polling (prevents hanging tests)
  - New contract events: FullTopicLog, DataOnlyLog, IndexedOnlyLog, burnGas()
5. Documentation (775 lines)
  - Test case specs for eth_feeHistory and eth_getLogs
  - RPC inventory and priority lists
6. Minor: assert!(matches!()) → assert_eq!() per coding guidelines

## (!) Handling breaking changes for existing chains

Make sure that evm genesis config has `"initial_base_fee": 0`. 

For all new rollups, evm genesis `initial_base_fee` should match exactly `INITIAL_BASE_FEE_PER_GAS[0]` from constants.toml  

In the future this value will become optional and rollup should fall back to constants.toml value

# More details below about how this kick started and applied.

-----

## Steps:

- Write draft of the prompt
- Refine prompt in Claude Web and ChatGPT
- Write down refined prompt after absorbing all comments
- Claude Code and Codex Both produce `docs/rpc_inventory.md` (codex) and `docs/rpc_inventory_claude.md` (claude) together with `rpc_prioirty`.
- Codex starts working on `eth_getLogs`, writing test cases in md file and them updating rust tests
- Claude starts working on `eth_feeHistory`, same way.
- Mid-through, separate instance of Codex reviews Claude's work and separate index of Claude review Codex's work. They mostly focus on priority and indidivual test cases
- Both original agents asked to absorb feedback and continue
- Another round of asking original agents to review test cases and focus on "real life" cases with actual data and veriyfing output.
- Then another round of asking original agents to verify that all cases in rust are covered

#### Context
Sovereign SDK rollup with EVM support. Beta testing revealed RPC inconsistencies. 

- Goal: RPC behavior indistinguishable from Ethereum L1 for end users. We want robust, high-ROI tests to ensure enabled Ethereum JSON-RPC endpoints behave correctly and return correct data.
- Primary focus: correctness around “latest/head” and “pending” behavior, especially for methods that accept a block selector/tag.

Scope rules
- This PR is TESTS + TEST HARNESS only. Do NOT fix production bugs as part of this task.
- If you detect a likely bug, clearly document it with a minimal reproduction from a test, but stop there.
- Error-code / error-shape conformance testing is OUT OF SCOPE for this PR (no dedicated invalid-param / error-shape tests).

Priority order for correctness
1) Semantic correctness (first priority)
2) Schema correctness (second priority)

# Codebase locations
- RPC implementation:  core EVM implementation in crates/module-system/module-implementations/sov-evm and some more RPC specicific in  crates/full-node/sov-ethereum
- Existing EVM tests: examples/demo-rollup/tests/evm
- Test utilities: crates/utils/sov-evm-test-utils

# Plan

## Phase 1: Research (output: prioritized endpoint list)
Identify the 10 most critical RPC endpoints by analyzing:
- What ethers.js, forge, viem, wagmi, foundry, hardhat actually call
- Failure modes that break UX (wallets, block explorers, dapps)
- Our current implementation status, put it into docs/rpc_inventory.md
- Produce a tiered list (P0/P1/P2/P3) and then a “Top 10 to test first” recommendation, if makes sense

Deliver as ranked list with rationale and write it into docs/rpc_priority.md.

## Phase 2: Deep-dive on one endpoint (I will select)
For the selected endpoint:
1. Document exact Ethereum spec behavior (params, responses, edge cases)
2. Pay special attention to block parameter handling: "latest", "pending", "finalized", "safe", specific numbers
3. Review existing test coverage in our codebase
4. Propose comprehensive test cases
5. Identify any contract deployments needed for testing

## Phase 3: Implement tests
Write tests following existing patterns in the codebase.

## Phase 4: Create extractable prompt
Produce a self-contained prompt I can use to start a fresh session for the next endpoint. It should include:
- The methodology we developed
- Test patterns established
- Any shared test infrastructure created
- Specific instructions for the target endpoint

This prompt must work standalone without referencing this conversation.

# Constraints
- Planning and testing only. Flag bugs, don't fix them.
- Prioritize by real-world tooling usage, not spec completeness
- Match existing test patterns

Anti-flakiness / test design constraints
- Tests must be deterministic: no sleeps, no dependence on wall-clock timing.
- Prefer explicit block production controls if available (or deterministic sequencing through the existing harness).
- Avoid asserting values that are expected to vary nondeterministically; assert stable invariants (relationships between fields, monotonic properties, equality where required).
- Keep tests fast and minimal while still meaningful (this is a high-ROI harness PR).

# References
- Spec: https://ethereum.github.io/execution-apis/
- Test examples: https://github.com/ethereum/execution-apis/tree/main/tests

----
----
----

## eth_feeHistory Test Coverage Complete

Results:
- 29 tests passing
- 3 tests ignored (documented bugs)

Bugs Found:

```
  ┌──────┬───────────────────────────────────────────────────────────────────────────────────┐
  │ Test │                                  Bug Description                                  │
  ├──────┼───────────────────────────────────────────────────────────────────────────────────┤
  │ TC14 │ Empty percentiles returns Some([[], []]) instead of None per spec                 │
  ├──────┼───────────────────────────────────────────────────────────────────────────────────┤
  │ TC27 │ Future block returns fabricated data (oldest_block=1001 when chain is at block 3) │
  ├──────┼───────────────────────────────────────────────────────────────────────────────────┤
  │ TC35 │ Base fee can drop to 0 due to saturating_sub in gas.rs (EIP-1559 violation)       │
  └──────┴───────────────────────────────────────────────────────────────────────────────────┘
```

```
 ┌────────────────────────────┬──────────┬────────────────────────────────┬───────────────────────────────────────────────────┬───────────────────────────────┐
  │             TC             │ Priority │          Description           │                     Rust Test                     │            Status             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ Block tag semantics        │          │                                │                                                   │                               │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC01                       │ P0       │ latest == pending              │ test_fee_history_latest_equals_pending            │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC02                       │ P0       │ finalized valid data           │ test_fee_history_finalized_tag                    │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC03                       │ P0       │ safe valid data                │ test_fee_history_safe_tag                         │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC04                       │ P1       │ earliest oldest_block=0        │ test_fee_history_earliest_tag                     │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC05                       │ P1       │ Number(N) oldest_block         │ test_fee_history_oldest_block_correctness         │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC06                       │ P2       │ finalized == safe              │ test_fee_history_finalized_equals_safe            │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ Parameter validation       │          │                                │                                                   │                               │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC07                       │ P0       │ blockCount=0 error             │ test_eth_fee_history_zero_blocks                  │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC08                       │ P1       │ blockCount>1024 capped         │ test_eth_fee_history_large_count_capped           │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC09                       │ P2       │ blockCount=1024 boundary       │ test_fee_history_block_count_1024_boundary        │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC10                       │ P0       │ percentile>100 error           │ test_fee_history_percentile_out_of_range_over_100 │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC11                       │ P1       │ percentile<0 error             │ —                                                 │ SKIP (client-side validation) │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC12                       │ P0       │ non-monotonic error            │ test_fee_history_percentiles_not_monotonic        │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC13                       │ P2       │ duplicate percentiles          │ test_fee_history_duplicate_percentiles            │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC14                       │ P2       │ empty percentiles → no reward  │ test_fee_history_empty_percentiles_no_reward      │ BUG                           │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ Response schema invariants │          │                                │                                                   │                               │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC15                       │ P0       │ base_fee.len == ratio.len+1    │ test_fee_history_array_length_invariants          │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC16                       │ P0       │ ratio.len == blockCount        │ test_fee_history_array_length_invariants          │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC17                       │ P1       │ oldest_block formula           │ test_fee_history_oldest_block_correctness         │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC18                       │ P1       │ reward dimensions              │ test_fee_history_array_length_invariants          │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC19                       │ P2       │ blob gas fields empty          │ test_fee_history_blob_gas_fields_empty            │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ Value correctness          │          │                                │                                                   │                               │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC20                       │ P0       │ gas_ratio in [0,1]             │ test_fee_history_gas_ratio_valid_range            │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC21                       │ P1       │ tx block ratio>0               │ test_fee_history_block_with_tx_nonzero_ratio      │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC22                       │ P1       │ base_fee non-negative          │ test_fee_history_percentile_boundaries            │ ✓ (implicit u128)             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC23                       │ P1       │ rewards are zero               │ test_eth_fee_history_with_reward_percentiles      │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC24                       │ P2       │ predicted next fee             │ test_fee_history_predicted_next_block_fee         │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ Edge cases                 │          │                                │                                                   │                               │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC25                       │ P1       │ blockCount=1 → 2 fees, 1 ratio │ test_eth_fee_history_single_block                 │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC26                       │ P2       │ beyond earliest graceful       │ test_fee_history_earliest_tag                     │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC27                       │ P3       │ future block handling          │ test_fee_history_future_block                     │ BUG                           │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC28                       │ P2       │ fractional percentiles         │ test_fee_history_fractional_percentiles           │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ State/history scenarios    │          │                                │                                                   │                               │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC29                       │ P0       │ empty blocks ratio=0           │ test_fee_history_empty_blocks_zero_ratio          │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC30                       │ P0       │ tx block ratio>0               │ test_fee_history_block_with_tx_nonzero_ratio      │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC31                       │ P1       │ multiple txs across blocks     │ test_fee_history_multiple_txs_across_blocks       │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC32                       │ P1       │ fee history progression        │ test_fee_history_progression                      │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC33                       │ P1       │ consistent query methods       │ test_fee_history_consistent_query_methods         │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC34                       │ P2       │ heavy gas usage                │ test_fee_history_heavy_gas_usage                  │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC35                       │ P2       │ mixed history                  │ test_fee_history_mixed_pattern                    │ ✓                             │
  ├────────────────────────────┼──────────┼────────────────────────────────┼───────────────────────────────────────────────────┼───────────────────────────────┤
  │ TC36                       │ P1       │ base fee stability             │ test_fee_history_base_fee_stability               │ BUG                           │
  └────────────────────────────┴──────────┴────────────────────────────────┴───────────────────────────────────────────────────┴───────────────────────────────┘
```

### eth_getLogs TC failures

  | TC# | Failing test | Bug signature (observed vs expected) |
  |---|---|---|
  | TC02 / TC03 | get_log_from_pending_block | Pending log has blockHash=None (expected synthetic block hash); blockHash filter can’t match pending logs (expected blockHash present  examples/demo-rollup/tests/evm/evm_logs.rs:1400 |
  | TC08 | get_logs_address_filter_single | Pending + address filter returns 0 logs (expected 2) examples/demo-rollup/tests/evm/evm_logs.rs:279 |
  | TC09 | get_logs_address_filter_multiple | Pending + address OR filter returns 0 logs (expected 5) examples/demo-rollup/tests/evm/evm_logs.rs:303 |
  | TC12 | get_logs_address_and_topic_intersection | Pending + address+topic filter returns 0 logs (expected 1) examples/demo-rollup/tests/evm/evm_logs.rs:408 |
  | TC13 | get_logs_pending_with_topic_filter | Pending + topic filter returns 0 logs (expected 5) examples/demo-rollup/tests/evm/evm_logs.rs:152 |
  | TC10 | get_logs_topic_or_semantics | Pending + topic OR returns 0 logs (expected 4) examples/demo-rollup/tests/evm/evm_logs.rs:339 |
  | TC11 | get_logs_topic_and_with_wildcard | Pending + topic+wildcard returns 0 logs (expected 1) examples/demo-rollup/tests/evm/evm_logs.rs:380 |
  | TC37 | get_logs_topic0_filters_event_signature | Pending + topic0 filter returns 0 logs (expected 2) examples/demo-rollup/tests/evm/evm_logs.rs:731 |
  | TC04 | get_logs_safe_finalized_exclude_pending | safe/finalized return empty list even though sealed logs exist (expected sealed logs) examples/demo-rollup/tests/evm/evm_logs.rs:457 |
  | TC20 | get_logs_eip1898_blockhash_object | fromBlock/toBlock EIP‑1898 object rejected: “invalid type: map, expected string” examples/demo-rollup/tests/evm/evm_logs.rs:1471 |

----
----

- [x] I have updated `CHANGELOG.md` with a new entry if my PR makes any breaking changes or fixes a bug. If my PR removes a feature or changes its behavior, I provide help for users on how to migrate to the new behavior.
- [x] I have carefully reviewed all my `Cargo.toml` changes before opening the PRs. (Are all new dependencies necessary? Is any module dependency leaked into the full-node (hint: it shouldn't)?)

## Linked Issues
- Related to #2271 

## Testing
No updates

## Docs
No updates

### CODE REVIEW

File: examples/demo-rollup/tests/evm/evm_logs.rs [hunk 1]
```diff
@@ -1,39 +1,1730 @@
...
+    let (rollup, client, _) = setup_with_simple_storage(0, EVM_EXTENSION).await;
+    let contract_address = client.alloy_deploy_contract().await;
+    rollup.wait_for_next_blocks(1).await;
+    rollup.pause_preferred_batches().await;
+
+    let sender = client.address();
+    let pending_filter = filter_for_tag(BlockNumberOrTag::Pending);
+
+    // Helper to build expected logs for a snapshot
+    let build_expected_at_snapshot = |block: &BlockContext,
```
Comment (bkolad) [inline examples/demo-rollup/tests/evm/evm_logs.rs]: This helper is overly verbose and fragile.
I think it should simply accept a `Vec<TxHash>` along with metadata about each transaction's logs. Then 
`Vec<ExpectedSimpleLog>`  can be produced in a generic loop.

File: examples/demo-rollup/tests/evm/evm_logs.rs [hunk 2]
```diff
@@ -1,39 +1,1730 @@
...
+        .await;
+
+    let tx_hashes = rollup_and_client
+        .produce_logs(nb_of_txs, nb_of_logs_per_tx, None)
+        .await;
+    let pending_block = latest_block_context(&rollup_and_client.client).await;
+    let sender = rollup_and_client.client.address();
+
+    let default_filter = Filter::new();
+    let latest_filter = filter_for_tag(BlockNumberOrTag::Latest);
```
Comment (bkolad) [inline examples/demo-rollup/tests/evm/evm_logs.rs:265]: Is it intended that we don't return logs from the pending block?

File: examples/demo-rollup/tests/evm/evm_logs.rs [hunk 3]
```diff
@@ -1,39 +1,1730 @@
...
+
+    rollup_and_client
+        .test_rollup
+        .pause_preferred_batches()
+        .await;
+
+    let tx_hashes = rollup_and_client
+        .produce_logs(nb_of_txs, nb_of_logs_per_tx, None)
+        .await;
+    let pending_block = latest_block_context(&rollup_and_client.client).await;
```
Comment (bkolad) [inline examples/demo-rollup/tests/evm/evm_logs.rs:261]: This should be called `latest_block`?

File: examples/demo-rollup/tests/evm/evm_logs.rs [hunk 4]
```diff
@@ -1,39 +1,1730 @@
...
+            log_count: nb_of_logs_per_tx as u64,
+        },
+        TxLogPlan {
+            tx_hash: tx_hashes[1],
+            log_count: nb_of_logs_per_tx as u64,
+        },
+    ];
+    let meta_map = build_tx_log_meta(&rollup_and_client.client, &plans).await;
+    let meta = meta_map
+        .get(&tx_hashes[0])
```
Comment (bkolad) [inline examples/demo-rollup/tests/evm/evm_logs.rs:334]: Why do we have two `TxLogPlan`, but in the loop, we only care about the first tx?

File: examples/demo-rollup/tests/evm/evm_logs.rs [hunk 5]
```diff
@@ -1,39 +1,1730 @@
...
+        .pause_preferred_batches()
+        .await;
+
+    let tx_hashes = rollup_and_client
+        .produce_logs(nb_of_txs, nb_of_logs_per_tx, None)
+        .await;
+    let pending_block = latest_block_context(&rollup_and_client.client).await;
+    let sender = rollup_and_client.client.address();
+
+    let topics = vec![U256::from(1).into(), U256::from(3).into()];
```
Comment (bkolad) [inline examples/demo-rollup/tests/evm/evm_logs.rs]: Does this mean that topic3 is either 1 or 3?  In that case lets call it topics_3?

File: examples/demo-rollup/tests/evm/evm_logs.rs [hunk 6]
```diff
@@ -1,39 +1,1730 @@
...
+            U256::ZERO,
+        );
+        assert!(filter.matches(log.inner.as_ref()));
+    }
+
+    Ok(())
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn get_logs_topic_or_semantics() -> anyhow::Result<()> {
```
Comment (bkolad) [inline examples/demo-rollup/tests/evm/evm_logs.rs:446]: What does `or_semantics` mean?

File: examples/demo-rollup/tests/evm/evm_logs.rs [hunk 7]
```diff
@@ -166,11 +1848,528 @@ async fn evm_test_get_logs_range_limit() {
...
+        data: U256,
+    },
+}
+
+struct ExpectedLogEntry {
+    meta: ExpectedLogMeta,
+    kind: ExpectedLogKind,
+}
+
+async fn block_context_by_number(client: &SimpleStorageClient, number: u64) -> BlockContext {
```
Comment (bkolad) [inline examples/demo-rollup/tests/evm/evm_logs.rs:1918]: Why do we need `BlockContext` can we just return the Block?

File: examples/demo-rollup/tests/evm/evm_logs.rs [hunk 8]
```diff
@@ -166,11 +1848,528 @@ async fn evm_test_get_logs_range_limit() {
...
+        .eth_get_block_by_number(Some(format!("0x{number:x}")))
+        .await;
+    BlockContext {
+        hash: block.header.hash,
+        number: block.number(),
+        timestamp: block.header.timestamp,
+    }
+}
+
+async fn latest_block_context(client: &SimpleStorageClient) -> BlockContext {
```
Comment (bkolad) [inline examples/demo-rollup/tests/evm/evm_logs.rs:1929]: Same

File: examples/demo-rollup/tests/evm/evm_logs.rs [hunk 9]
```diff
@@ -166,11 +1848,528 @@ async fn evm_test_get_logs_range_limit() {
...
+        by_block.entry(block_number).or_default().push(PlanReceipt {
+            plan: *plan,
+            block_number,
+            block_hash,
+            tx_index,
+        });
+    }
+
+    let mut meta_map = HashMap::new();
+    for (block_number, mut entries) in by_block {
```
Comment (bkolad) [inline examples/demo-rollup/tests/evm/evm_logs.rs]: It looks like `PlanReceipt` could be removed, and we could do everything in a single loop. On a high level, it looks like this method just returns Reciept with block timestamp.

File: crates/module-system/module-implementations/sov-evm/src/rpc/fee_history.rs [hunk 10]
```diff
@@ -176,12 +188,45 @@ where
...
-            gas_limit,
-            conservative_parent_gas_usage,
-            Amount::from(*last_base_fee),
-        );
-        base_fees.push(next_gas_price.0.try_into().unwrap_or(u64::MAX));
+
+        // Validate that the EVM block number aligns with the current rollup height.
+        // We allow a +1 offset to account for the pending block, but anything beyond that
+        // indicates a mapping mismatch between EVM blocks and rollup heights.
+        let current_rollup_height = self.chain_state_module.rollup_height(state)?;
```
Comment (preston-evans98) [inline crates/module-system/module-implementations/sov-evm/src/rpc/fee_history.rs]: nit: We can get the current height from the accessor without querying state `state.rollup_height_to_access`

**Review by preston-evans98 — APPROVED** [review]

File: examples/demo-rollup/tests/evm/evm_logs.rs (hunk 7 above)
Comment (citizen-stig) [reply examples/demo-rollup/tests/evm/evm_logs.rs:1918]: We can, but BlockContext is a smaller struct with only 3 fields, so we know exactly what tests can do with, where full block will have more data, which mean:

* More verbose accesses in test
* Test can pontentially start relying on more fields, making tests more convoluted.

I don't have strong opinion on this one and happy to change, but I think net result will be about the same. I can implement `From<Block> for BlockContext` to trim some lines, but considering size of the whole it is going to be miniscule.

If name BlockContext is confusing, we can work on that.

File: examples/demo-rollup/tests/evm/evm_logs.rs (hunk 9 above)
Comment (citizen-stig) [reply examples/demo-rollup/tests/evm/evm_logs.rs]: Addressed that

File: examples/demo-rollup/tests/evm/evm_logs.rs (hunk 6 above)
Comment (citizen-stig) [reply examples/demo-rollup/tests/evm/evm_logs.rs:446]: Added a comment (not displaed here, for some reason.

File: examples/demo-rollup/tests/evm/evm_logs.rs (hunk 1 above)
Comment (citizen-stig) [reply examples/demo-rollup/tests/evm/evm_logs.rs]: Addressed. Now it is more declarative for each tx, and helper just simply transforms the data.

File: examples/demo-rollup/tests/evm/evm_logs.rs (hunk 5 above)
Comment (citizen-stig) [reply examples/demo-rollup/tests/evm/evm_logs.rs]: Renamed variable

File: examples/demo-rollup/tests/evm/evm_logs.rs (hunk 2 above)
Comment (citizen-stig) [reply examples/demo-rollup/tests/evm/evm_logs.rs:265]: We do. Our semantics is special. We have `pending == latest`

File: examples/demo-rollup/tests/evm/evm_logs.rs (hunk 3 above)
Comment (citizen-stig) [reply examples/demo-rollup/tests/evm/evm_logs.rs:261]: Good point. Addressed.

File: examples/demo-rollup/tests/evm/evm_logs.rs (hunk 4 above)
Comment (citizen-stig) [reply examples/demo-rollup/tests/evm/evm_logs.rs:334]: Good point, addressed

**Review by bkolad — APPROVED** [review]

File: crates/module-system/module-implementations/sov-evm/src/rpc/fee_history.rs (hunk 10 above)
Comment (citizen-stig) [reply crates/module-system/module-implementations/sov-evm/src/rpc/fee_history.rs]: Thanks for the tip!
